{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c8a80b8f-ec04-4f0f-b12e-4d80d19aedd2",
   "metadata": {},
   "source": [
    "# Open Reading Frames\n",
    "https://rosalind.info/problems/orf/b"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cfbeaa03",
   "metadata": {},
   "source": [
    "## My solution"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0318115e-6aae-4184-867d-49ef2403cbc7",
   "metadata": {},
   "source": [
    "### Practice "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "56564483-7118-4850-87be-65afc2424f75",
   "metadata": {},
   "outputs": [],
   "source": [
    "from util import read_file_into_lines\n",
    "from util import text_to_fasta\n",
    "from util import codon_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7c2cd39f-a8e9-4457-a20b-793cd43547e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "s = \"AGCCATGTAGCTAACTCAGGTTACATGGGGATGACCCCGCGACTTGGATTAGAGTCTCTTTTGGAATAAGCCTGAATGATCCGAGTAGCATCTCAG\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "02b23091-3440-498c-8a8f-c6a674e40588",
   "metadata": {},
   "outputs": [],
   "source": [
    "counter = 0\n",
    "peptide = \"\"\n",
    "start_codon = \"ATG\"\n",
    "stop_codons = [\"TAG\", \"TGA\", \"TAA\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6f005a01-9f64-4626-a08a-3647d8bc6be6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGC\n",
      "GCC\n",
      "CCA\n",
      "CAT\n",
      "start codon ATG\n",
      "TGT\n",
      "GTA\n",
      "end codon TAG\n"
     ]
    }
   ],
   "source": [
    "# this did not work but it was a preliminary thinking\n",
    "for i in range(len(s)):\n",
    "    if s[i:i+3] == start_codon:\n",
    "        print(\"start codon\", s[i:i+3])       \n",
    "    elif s[i:i+3] in stop_codons:\n",
    "        print(\"end codon\", s[i:i+3])\n",
    "        peptide += s[i:i+3]\n",
    "        break\n",
    "    else: \n",
    "        print(s[i:i+3])\n",
    "        peptide += s[i:i+3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "96176e43-cd55-4c8f-bc6a-c5260bec08df",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_candidate_string(dna_string):\n",
    "    current_list = []\n",
    "    for i in range(len(dna_string)):\n",
    "        if dna_string[i:i+3] == \"ATG\":\n",
    "            candidate_string = \"\"\n",
    "            for j in range(i, len(dna_string), 3):\n",
    "                codon = dna_string[j:j+3]\n",
    "                if codon in stop_codons:\n",
    "                    current_list.append(candidate_string)\n",
    "                    break\n",
    "                else: \n",
    "                    candidate_string += dna_string[j:j+3]\n",
    "    return(current_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f1a4e3d7-2c5b-493b-a781-65757ae86164",
   "metadata": {},
   "outputs": [],
   "source": [
    "candidate_list = find_candidate_string(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bc150fa2-ac0a-4453-86d7-68a6cc49fccb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ATG',\n",
       " 'ATGGGGATGACCCCGCGACTTGGATTAGAGTCTCTTTTGGAA',\n",
       " 'ATGACCCCGCGACTTGGATTAGAGTCTCTTTTGGAA']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "candidate_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7a002f52-681f-4903-8ac6-9119ecf7f856",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ATG', 'ATGGGGATGACCCCGCGACTTGGATTAGAGTCTCTTTTGGAA', 'ATGACCCCGCGACTTGGATTAGAGTCTCTTTTGGAA', 'ATGCTACTCGGATCATTCAGGCTTATTCCAAAAGAGACTCTAATCCAAGTCGCGGGGTCATCCCCATGTAACCTGAGT', 'ATG']\n"
     ]
    }
   ],
   "source": [
    "sc = s[::-1].replace(\"A\",\"t\").replace(\"T\", \"a\").replace(\"C\", \"g\").replace(\"G\", \"c\").upper()\n",
    "\n",
    "candidate_list.extend(find_candidate_string(sc))\n",
    "print(candidate_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1e91df19-a94f-405c-b657-476a56cae7a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def translate_dna_rna(candidate_list):\n",
    "    rna_list = []\n",
    "    for i in candidate_list:\n",
    "        rna = \"\"\n",
    "        for nucl in i:\n",
    "            if nucl == \"T\":\n",
    "                rna = rna + \"U\"\n",
    "            else:\n",
    "                rna = rna + nucl\n",
    "        rna_list.append(rna)\n",
    "    return(rna_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "4adc9783-9dc7-4da5-ab51-346d3de2e186",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_list = translate_dna_rna(candidate_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a5284c48",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AUG',\n",
       " 'AUGGGGAUGACCCCGCGACUUGGAUUAGAGUCUCUUUUGGAA',\n",
       " 'AUGACCCCGCGACUUGGAUUAGAGUCUCUUUUGGAA',\n",
       " 'AUGCUACUCGGAUCAUUCAGGCUUAUUCCAAAAGAGACUCUAAUCCAAGUCGCGGGGUCAUCCCCAUGUAACCUGAGU',\n",
       " 'AUG']"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "04482067-fc40-4591-96ab-7eb35d7dc379",
   "metadata": {},
   "outputs": [],
   "source": [
    "def from_rna_to_protein(rna_list):\n",
    "    protein_list = []\n",
    "    for i in rna_list:\n",
    "        counter = 0 \n",
    "        aa = \"\"\n",
    "        for codon in i:\n",
    "            triplet = i[counter:counter + 3] \n",
    "            if triplet == '': \n",
    "                break\n",
    "            else:\n",
    "                aa = aa + codon_dict[triplet] \n",
    "                counter = counter + 3\n",
    "        protein_list.append(aa)\n",
    "    return(protein_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "951691b3-4f0e-4803-8a65-95a84dba2410",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['M', 'MGMTPRLGLESLLE', 'MTPRLGLESLLE', 'MLLGSFRLIPKETLIQVAGSSPCNLS', 'M']"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from_rna_to_protein(rna_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "835ae90e-99f9-4b64-91f0-e70f2c16d75c",
   "metadata": {},
   "source": [
    "### All together \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "1f6353b8-d47a-417d-884b-944efe574aa3",
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_file = read_file_into_lines(\"rosalind_orf.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "66004760-47f1-49a3-a8d1-06d6f6c72621",
   "metadata": {},
   "outputs": [],
   "source": [
    "#now we can put all the functions together into one program so that we can input a rasta file and have a resulting list of protein chains\n",
    "\n",
    "def fasta_to_protein(fasta_file):\n",
    "    dna_list = text_to_fasta(fasta_file)\n",
    "    for dna_string in dna_list.values(): \n",
    "        candidate_list = find_candidate_string(dna_string)\n",
    "        sc = dna_string[::-1].replace(\"A\",\"t\").replace(\"T\", \"a\").replace(\"C\", \"g\").replace(\"G\", \"c\").upper()\n",
    "        candidate_list.extend(find_candidate_string(sc))\n",
    "    rna_list = translate_dna_rna(candidate_list)\n",
    "    protein_list = from_rna_to_protein(rna_list)\n",
    "    return protein_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "5caa7f0f-0063-4d71-b98a-ee4615317704",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MPSLKSYGEQK\n",
      "MRSH\n",
      "MLAAERARRSSDTSTANCVLPSC\n",
      "MSLLHAVNRNLTSPAPLILDPDHELSAGPLHIGKITAEVSVQPRVG\n",
      "M\n",
      "MIILGGGGIQKGQNFVYRPIYHHCD\n",
      "MHYPPLLAQEKFPTSSGSRLAFP\n",
      "MSLLITITFCVTPLGSEPMVVVRPVTHVVAR\n",
      "MVVVRPVTHVVAR\n",
      "MLGGSH\n",
      "MHPLPIAKVQDGLGFVLV\n",
      "MSPSMNRTTMRLITMVIDRTINKVLTFLNSSAT\n",
      "MNRTTMRLITMVIDRTINKVLTFLNSSAT\n",
      "MRLITMVIDRTINKVLTFLNSSAT\n",
      "MVIDRTINKVLTFLNSSAT\n",
      "MSYVIANVRDATMARCS\n",
      "MRQWRDVAKQRLFERLAKRSKSRCLATFFLANSRLNAYLGCNLTNVKRSSRQFVVGV\n",
      "MARCS\n",
      "M\n",
      "M\n",
      "MCPATRLPV\n",
      "ML\n",
      "MGTHPR\n"
     ]
    }
   ],
   "source": [
    "final_solution = fasta_to_protein(fasta_file)\n",
    "for protein in final_solution:\n",
    "    print(protein)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee4aad53-f96c-4ee0-9fbd-f56db8de493f",
   "metadata": {},
   "source": [
    "### How do I avoid having double entries for proteins that already exist? "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62393a56-1264-43a7-b787-e1da2a0c24f2",
   "metadata": {},
   "source": [
    "## In class solution "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "248afe69",
   "metadata": {},
   "source": [
    "**Given**: A DNA sequence in FASTA format\n",
    "\n",
    "**Asked:** all possible peptides that can be translated from it\n",
    "\n",
    "### The Plan\n",
    "\n",
    "- [x] transcribe the DNA to RNA (RNA exercise)\n",
    "- [x] find all reading frames $\\rightarrow$ three forward, three reverse $\\rightarrow$ REVC exercise\n",
    "- [x] find valid ORFs in an RNA sequence <font color=\"red\">!!!!</font>\n",
    "- [x] translate the mRNAs to peptides (PROT exercise)\n",
    "- [x] make sure there are no duplicate protein sequences!\n",
    "\n",
    "Ideas:\n",
    "\n",
    "- much like in PROT, we might need to work with triplets\n",
    "- whatever we do for a single reading frame, we will have to repeat 5x more $\\rightarrow$ functions?\n",
    "    - function that takes RNA sequence and produces all valid peptides"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "559b53bb",
   "metadata": {},
   "source": [
    "### Set up environment\n",
    "\n",
    "- [x] load helper functions\n",
    "- [x] get testing data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "9888974d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from util import read_file_into_lines, read_fasta, reverse_complement\n",
    "from util import codon_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "268d913d",
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_file = read_fasta('orf_trial.txt')\n",
    "fasta = list(fasta_file.values())[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "4db16411",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AGCCATGTAGCTAACTCAGGTTACATGGGGATGACCCCGCGACTTGGATTAGAGTCTCTTTTGGAATAAGCCTGAATGATCCGAGTAGCATCTCAG'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a70cd5a",
   "metadata": {},
   "source": [
    "### 1. transcribe to RNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "826a0e65",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna = fasta.replace('T', 'U')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e8f33c9",
   "metadata": {},
   "source": [
    "### 2. find all reading frames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "28100a3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_reading_frames(rna):\n",
    "    # three forward, three backward reading frames\n",
    "    rf1 = rna\n",
    "    rf2 = rna[1:]\n",
    "    rf3 = rna[2:]\n",
    "\n",
    "    revc = reverse_complement(rna)\n",
    "    rf4 = revc\n",
    "    rf5 = revc[1:]\n",
    "    rf6 = revc[2:]\n",
    "    return [rf1, rf2, rf3, rf4, rf5, rf6]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "eb2d1328",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AGCCAUGUAGCUAACUCAGGUUACAUGGGGAUGACCCCGCGACUUGGAUUAGAGUCUCUUUUGGAAUAAGCCUGAAUGAUCCGAGUAGCAUCUCAG',\n",
       " 'GCCAUGUAGCUAACUCAGGUUACAUGGGGAUGACCCCGCGACUUGGAUUAGAGUCUCUUUUGGAAUAAGCCUGAAUGAUCCGAGUAGCAUCUCAG',\n",
       " 'CCAUGUAGCUAACUCAGGUUACAUGGGGAUGACCCCGCGACUUGGAUUAGAGUCUCUUUUGGAAUAAGCCUGAAUGAUCCGAGUAGCAUCUCAG',\n",
       " 'CUGAGAUGCUACUCGGAUCAUUCAGGCUUAUUCCAAAAGAGACUCUAAUCCAAGUCGCGGGGUCAUCCCCAUGUAACCUGAGUUAGCUACAUGGCU',\n",
       " 'UGAGAUGCUACUCGGAUCAUUCAGGCUUAUUCCAAAAGAGACUCUAAUCCAAGUCGCGGGGUCAUCCCCAUGUAACCUGAGUUAGCUACAUGGCU',\n",
       " 'GAGAUGCUACUCGGAUCAUUCAGGCUUAUUCCAAAAGAGACUCUAAUCCAAGUCGCGGGGUCAUCCCCAUGUAACCUGAGUUAGCUACAUGGCU']"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "frames = get_reading_frames(rna)\n",
    "frames"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cdfddb65",
   "metadata": {},
   "source": [
    "### 3. finding ORFs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "d162c2b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# just to practice, use RF1\n",
    "rf = frames[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37f3730f",
   "metadata": {},
   "source": [
    "I notice that I cannot solve this in one go; the problem are overlapping peptides. What is the\n",
    "problem  about them? They share a stop codon, but have different start codons $\\rightarrow$ there\n",
    "is no solution with just one for loop.\n",
    "\n",
    "My idea:\n",
    "\n",
    "- [x] find all start codons\n",
    "- [x] given a start codon, find the correspodnign ORF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "a6d444b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_AUG(rf):\n",
    "    starting_positions = []\n",
    "    for i in range(0, len(rf), 3):\n",
    "        codon = rf[i:i+3]\n",
    "        if codon == 'AUG':\n",
    "            starting_positions.append(i)\n",
    "    return starting_positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "b8568339",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[24, 30, 75]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "starts = find_AUG(rf)\n",
    "starts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "217562d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_ORF_and_translate(rf, start, codons):\n",
    "    peptide = ''\n",
    "    for i in range(start, len(rf), 3):\n",
    "        codon = rf[i:i+3]\n",
    "        if len(codon) < 3:\n",
    "            continue\n",
    "        if codons[codon] is None:\n",
    "            return peptide\n",
    "        peptide += codons[codon]\n",
    "    return None"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9714c3af",
   "metadata": {},
   "source": [
    "negative test: this sequence contains a start codon but no end codon; our function should return\n",
    "None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "28d2b6a3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "None\n"
     ]
    }
   ],
   "source": [
    "neg = 'AUGGGGAUGACCCCGCGACUUGGAUUAGAGUCUCUUUUGGAA'\n",
    "peptide = find_ORF_and_translate(neg, 0, codon_dict)\n",
    "print(peptide)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "4c97e8b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_all_proteins(rf, codons):\n",
    "    starts = find_AUG(rf)\n",
    "    proteins = []\n",
    "    for start in starts:\n",
    "        peptide = find_ORF_and_translate(rf, start, codons)\n",
    "        if peptide is not None:\n",
    "            proteins.append(peptide)\n",
    "    return proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "1cb226c2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "find_all_proteins(frames[0], codon_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "805bd746",
   "metadata": {},
   "source": [
    "### Putting it all together:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "8137b761",
   "metadata": {},
   "outputs": [],
   "source": [
    "def all_possible_proteins(fasta_path, codons):\n",
    "    # 1. read the file\n",
    "    fasta_file = read_fasta(fasta_path)\n",
    "    fasta = list(fasta_file.values())[0]\n",
    "    # 2. transcribe to RNA\n",
    "    rna = fasta.replace('T', 'U')\n",
    "    # 3. get reading frames\n",
    "    frames = get_reading_frames(rna)\n",
    "    proteins = []\n",
    "    for frame in frames:\n",
    "        to_add = find_all_proteins(frame, codons)\n",
    "        # make sure the new proteins are not\n",
    "        # already in our list\n",
    "        for prot in to_add:\n",
    "            if prot not in proteins:\n",
    "                proteins.extend([prot])\n",
    "\n",
    "    print('\\n'.join(proteins))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4afb2a88",
   "metadata": {},
   "source": [
    "### Solve example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "1c24ad6a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "all_possible_proteins('orf_trial.txt', codon_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f565d66f",
   "metadata": {},
   "source": [
    "### Solve exercise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "c2f1071b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "all_possible_proteins('rosalind_orf.txt', codon_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70e1792f",
   "metadata": {},
   "source": [
    "For some reason, this is not working for me. Might be useful to try with Niko's notebook and util.py file."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
